aten/src/ATen/templates/TensorBody.h

```diff
@@ -800,3 +793,113 @@ static inline DispatchKey legacyExtractDispatchKey(const Tensor& t) {
 }
 
 } // namespace at
+
+namespace c10 {
+  template<> class optional<at::Tensor> {
+    public:
+      using value_type = at::Tensor;
+
+      constexpr optional() noexcept
+      : optional(nullopt) {}
+
+      constexpr optional(nullopt_t) noexcept
+      : tensorOrUndefined_() {}
+
+      optional(const optional&) = default;
+      optional(optional&&) noexcept = default;
+      optional& operator=(nullopt_t) noexcept {
+        tensorOrUndefined_.reset();
+        return *this;
+      }
+      optional& operator=(const optional&) = default;
+      optional& operator=(optional&&) noexcept = default;
+
+      const at::Tensor* operator->() const {
+        return &tensorOrUndefined_;
+      }
+
+      at::Tensor* operator->() {
+        return &tensorOrUndefined_;
+      }
+
+      const at::Tensor& operator*() const& {
+        return tensorOrUndefined_;
+      }
+
+      at::Tensor& operator*() & {
+        return tensorOrUndefined_;
+      }
+
+      const at::Tensor&& operator*() const&& {
+        return std::move(tensorOrUndefined_);
+      }
+
+      at::Tensor&& operator*() && {
+        return std::move(tensorOrUndefined_);
+      }
+
+      template <class U = value_type, std::enable_if_t<
+          std::is_constructible<at::Tensor, U&&>::value
+          && !std::is_same<typename std::decay<U>::type, in_place_t>::value
+          && !std::is_same<typename std::decay<U>::type, optional<at::Tensor>>::value, int> = 0>
+      optional(U&& value)
+      : tensorOrUndefined_(at::Tensor(std::move(value)).getIntrusivePtr()) {
```

Comment:
why the `getIntrusivePtr()` call? it seems that constructing directly (`tensorOrUndefined_(std::move(value))`) would be more efficient. Writing it this way emits a refcount decrement for the temporary at::Tensor().

Reply:
Hmm, I can't seem to get the decrement to go away. Not sure where I've gone wrong, but I guess I stand corrected.

Reply:
I meant to retract this comment last week, but apparently my retraction didn't get posted. I couldn't get the decrement to go away by removing the `getIntrusivePtr()` call, so I stand corrected.